packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts

```diff
@@ -0,0 +1,186 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+import { ExtractionClient, ExtractorState, MappingsClient, REPORTING_BASE_PATH, ReportsClient } from "@itwin/insights-client";
+import type { ReportMapping } from "@itwin/insights-client";
+import { generateUrl, handleError } from "./utils";
+import type { CreateTypeFromInterface } from "./utils";
+import type {
+  GetSingleIModelParams,
+} from "@itwin/imodels-client-management";
+import { Constants, IModelsClient } from "@itwin/imodels-client-management";
+import { AccessTokenAdapter } from "@itwin/imodels-access-frontend";
+import type { ReportsApiConfig } from "../context/ReportsApiConfigContext";
+import { ExtractionStates } from "./ExtractionStatus";
+
+export type ReportMappingType = CreateTypeFromInterface<ReportMapping>;
+
+export type ReportMappingAndMapping = ReportMappingType & {
+  mappingName: string;
+  mappingDescription: string;
+  iModelName: string;
+};
+
+export default class BulkExtractor {
+  private _reportRunIds = new Map<string, string[]>();
+  private _reportsClientApi: ReportsClient;
+  private _mappingsClientApi: MappingsClient;
+  private _iModelsClient: IModelsClient;
+  private _extractionClientApi: ExtractionClient;
+  private _accessToken: Promise<string>;
```

Comment:
This should not be a promise. This should be a function which returns a promise.
```
private _accessToken: () => Promise<string>;
```

Reply:
Your implementation would be fine if AccessTokens would never expire, but they do expire and it would break this component.

Reply:
`_accesstToken` type changed to `() => Promise<string>;`